[Validator] Add ConstraintViolationBuilder methods: fromViolation(), setPath(), getViolation()

rela589n opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | https://github.com/symfony/symfony/issues/58029#issuecomment-2722367527
| License       | MIT

This PR: (1) adds the ability to create constraint violation builder from an existing violation (static factory method: `ConstraintViolationBuilder::fromViolation($violation)`) so that it can be adjusted in the builder, in particular (2) `setPath()` method, and finally retrieve new violation with (3) `getViolation()` method.

adrianrudnik committed:
[Scheduler] Throw error on duplicate schedule provider service registration on the schedule name

wkania committed:
[Form] Add input=date_point to DateTimeType, DateType and TimeType

crydotsnake committed:
[Dotenv] improve documentation for dotenv component

rela589n committed:
[Validator] Add ConstraintViolationBuilder methods: fromViolation(), setPath(), getViolation()

alamirault committed:
Replace get_class() calls by ::class

nicolas-grekas committed:
minor #60581 Replace `get_class()` calls by `::class` (alamirault)

This PR was merged into the 7.4 branch.

Discussion
----------

Replace `get_class()` calls by `::class`

| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

Replace `get_class()` by `::class`
It was already done in past in https://github.com/symfony/symfony/pull/47401

Commits
-------

e0a602bfc57 Replace get_class() calls by ::class

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilderInterface.php:
```diff
@@ -84,13 +89,13 @@ public function setInvalidValue(mixed $invalidValue): static;
      * @see \Symfony\Contracts\Translation\TranslatorInterface::trans()
      */
-    public function setPlural(int $number): static;
+    public function setPlural(?int $number): static;
```
this is a BC break
looking at the decoration test case, this might not be needed - instead, don't call setPlural if getPlural returns null

rela589n replied:
you're absolutely right, thank you for pointing this out
for some reason I didn't think about decoration :face_with_diagonal_mouth:

rela589n replied:
I've changed the interface 

What about `ConstarintViolationBuilder` itself?

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -90,7 +120,7 @@ public function setInvalidValue(mixed $invalidValue): static
     }
 
-    public function setPlural(int $number): static
+    public function setPlural(?int $number): static
```
see my concern about this change in the interface

rela589n replied:
changed the interface

btw, from my point question arises: if at any point we'd like to extend interface, how is it handled? 
is this change released on a new major version as bc break? is there any deprecation of "not using new type"?

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -27,24 +28,52 @@
 class ConstraintViolationBuilder implements ConstraintViolationBuilderInterface
 {
     private string $propertyPath;
+    private ?string $message = null;
```
no need to make this nullable:
```suggestion
    private string $translatedMessage;
```

rela589n replied:
is it the point that it will be eventually initialized?

right now this is only initialized as `$this->message ??= $this->translateMessage()`

any access to this property prior to that place will throw an exception

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -27,24 +28,52 @@
+        private ?ConstraintViolationListInterface $violations,
         private ?Constraint $constraint,
-        private string|\Stringable $message,
+        private string|\Stringable $messageTemplate,
```
let's keep the previous name

rela589n replied:
I was motivated by using the same naming as in `ConstraintViolation`, as these properties are named there as `$message` and `$messageTemplate`.

Naming them as `$message` and `$translatedMessage` here would confuse the picture, since `$message` means different things. Though, if you'd like them to be named this way, I will make a change.

Frankly speaking I myself thought about adding `$translatedMessage`, and letting the original `$message` alone, because the rename might introduce a BC break, as changing parameter name from `message` to `messageTemplate` would break the invocation code if it passes it as a named parameter.

Though, I've checked BC promise, and it says:

> [10] Parameter names are only covered by the compatibility promise for constructors of Attribute classes. Using PHP named arguments might break your code when upgrading to newer Symfony versions.
>
> [11] Only optional argument(s) of a constructor at last position may be added. 

So, keeping it unified looks better to me, but in any case, feel free to ask what you will

nicolas-grekas reviewed (commented):
Please add a line to the changelog of the component also

nicolas-grekas committed:
minor #60570 [Dotenv] improve documentation for dotenv component (crydotsnake)

This PR was squashed before being merged into the 7.4 branch.

Discussion
----------

[Dotenv] improve documentation for dotenv component

| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Improves the documentation for the Symfony dotenv component :)

Commits
-------

adfc7e97bcc [Dotenv] improve documentation for dotenv component

nicolas-grekas committed:
feature #60315 [Form] Add `input=date_point` to `DateTimeType`, `DateType` and `TimeType` (wkania)

This PR was merged into the 7.4 branch.

Discussion
----------

[Form] Add `input=date_point` to `DateTimeType`, `DateType` and `TimeType`

| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        |
| License       | MIT

Based on [datetime_immutable](https://symfony.com/blog/new-in-symfony-4-1-added-support-for-immutable-dates-in-forms).

After [DatePointType](https://github.com/symfony/symfony/pull/59900) and [DatePointDateType](https://github.com/symfony/symfony/pull/60237), it would be great to use Forms without needing to transform values into the DatePoint type manually.

```
use Symfony\Component\Form\Extension\Core\Type\DateType;
use Symfony\Component\Form\Extension\Core\Type\DateTimeType;
use Symfony\Component\Form\Extension\Core\Type\TimeType;
use Symfony\Component\Form\Extension\Core\Type\BirthdayType;

$builder->add('from', DateType::class, [
    'input' => 'date_point',
]);
$builder->add('from', DateTimeType::class, [
    'input' => 'date_point',
]);
$builder->add('from', TimeType::class, [
    'input' => 'date_point',
]);
$builder->add('from', BirthdayType::class, [
    'input' => 'date_point',
]);
```

Alternative: Make symfony/clock a hard requirement and refactor the existing DateTimeImmutableToDateTimeTransformer to return a DatePoint instead. This should not introduce any breaking changes.

Commits
-------

f1160d6617f [Form] Add input=date_point to DateTimeType, DateType and TimeType

nicolas-grekas committed:
feature #60253 [Scheduler] Throw error on duplicate schedule provider service registration on the schedule name (adrianrudnik)

This PR was squashed before being merged into the 7.4 branch.

Discussion
----------

[Scheduler] Throw error on duplicate schedule provider service registration on the schedule name

| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | See below
| License       | MIT

The way the schedule providers work, either by tagging a service or using the [AsSchedule](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Scheduler/Attribute/AsSchedule.php) attribute, can lead to a scenario where multiple [ScheduleProviders](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Scheduler/ScheduleProviderInterface.php) register for the same schedule name (e.g. `default`).

The problem arises when the [CompilerPass](https://github.com/symfony/symfony/blob/29da4f53d4a5b6a48ae8b43330e764a86ad7a85b/src/Symfony/Component/Scheduler/DependencyInjection/AddScheduleMessengerPass.php#L49) simply overwrites the previously registered one without throwing a warning, a notice or anything else. The problem would be that a full ScheduleProvider would simply not register and therefore not run.

I decided to use the InvalidArgumentException from DI, as I could not get a trigger_error on E_USER_WARNING to show anything in a `dev` environment, in case I miss a scenario where this would actually be a desirable use case.

Not sure if the test is complete enough for this scenario. I tried to base it on others found in HttpKernel.

Also not sure if this falls under "feature" or "bug fix", so I marked it as a feature above.

Commits
-------

8548aac4af8 [Scheduler] Throw error on duplicate schedule provider service registration on the schedule name

rela589n committed:
[Validator] Fix BC break about ConstraintViolationBuilderInterface::setPlural()

xabbuh commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -27,24 +28,52 @@
 
     public function __construct(
-        private ConstraintViolationList $violations,
+        private ?ConstraintViolationListInterface $violations,
```
Making this property nullable doesn't look good to me as you would now have to be aware of the inner state of the `ConstraintViolationBuilder` class to decide whether or not you can call `addViolation()`.

rela589n replied:
@xabbuh , what do you propose instead?

rela589n replied:
> to decide whether or not you can call

how often is it necessary to "decide" in the way you say it?

afaik, `addViolation` is only called in custom validators that create violation builder using execution context, and it itself provides `ConstraintViolationList`

So what is the point in "deciding"?

xabbuh replied:
Take a look at this arbitrary example taken from the Symfony code base (from the `ChoiceValidator`):

```php
$this->context->buildViolation($constraint->maxMessage)
    ->setParameter('{{ limit }}', $constraint->max)
    ->setPlural((int) $constraint->max)
    ->setCode(Choice::TOO_MANY_ERROR)
    ->addViolation();
```

You wouldn't know here if the `ConstraintViolationBuilderInterface` implementation returned by `buildViolation()` would throw or not.

rela589n replied:
Sorry, but what do you mean that we wouldn't know if it will throw or not?
It will never throw

xabbuh replied:
As a consumer of the builder one cannot know whether or not everything is set inside the object being retrieved here so you wouldn't know if `addViolation()` throws because no list of violations was set.

rela589n replied:
Yes, but are you living in the constant fear of all the factors you can't control? What's the matter of worrying about it?
It's not the client code's business to create constraint violation builder, neither is it its business to know if it's throwing inside or not. Be relaxed, it never throws.

xabbuh replied:
That's simply not true. The method can throw with your implementation and one would have to account for that.

rela589n replied:
There is some precondition for it to throw - that violations property must be null.
But it is never null in the scenario which you describe.

nicolas-grekas replied:
>  But it is never null in the scenario which you describe.

Before, it was guaranteed.
With this change, it's going to work by convention, not by contracts anymore.
That's what @xabbuh means here IIUC.

rela589n replied:
Do you think that in the future someone might stop passing violations to constraint violation builder?

nicolas-grekas replied:
That's a core design principle: not allowing clumsy behavior by making the design forbid them. Weakening means increasing the risk for others, while the benefit we'd get is to cover a niche use case that can be implemented another way.

rela589n replied:
Actually I do not like the idea of using constraint violation builder both for building and adding violations to the list. Though, in current circumstances when there's exactly one builder that is used both to build and to add violations, what's the best way? A separate class that entirely duplicates the current builder but `addViolation()` method?

rela589n replied:
I agree that building and adding violations are two different purposes, but isn't it what already happens in the class?

rela589n replied:
@nicolas-grekas , @xabbuh , what do you think about the fact that this is already so?

xabbuh replied:
Maybe we would do things differently when we were to build the component again and split responsibility here. But given the impact that would have if we split the current implementation this is not going to happen.

rela589n replied:
Yes, and I believe that these two must be kept in one. And if one axis is to be extended, the other must be nullable

rela589n replied:
@xabbuh , can you concede with this?

xabbuh commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -27,24 +28,52 @@
+            ->setCause($violation->getCause());
+    }
+
+    public function setPath(string $path): static
```
having this method is probably confusing for users of this class given that we already have an `atPath()` method and by looking at the names it's not obvious how they differ.

rela589n replied:
In most programming languages, setter methods (which modify private variables) are conventionally named by starting with "set" followed by the variable name, with the first letter of the variable name capitalized. For example, if a variable is named firstName, the setter would be setFirstName()

rela589n replied:
I agree that building and adding violations are two different purposes, but isn't it what already happens in the class?

xabbuh commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilderInterface.php:
```diff
@@ -11,15 +11,20 @@
  *
  * @author Bernhard Schussek <bschussek@gmail.com>
+ *
+ * @method ConstraintViolationInterface getViolation()
```
Adding this method adds a complete new meaning to the interface. I don't feel that it's a good idea having the same interface serve two different purposes.

rela589n replied:
Guys I sincerely believe that this is the only way that can make the existing code applicable for the features that involve violation building

xabbuh replied:
Here is one last answer from my side: The changes that you proposed will not be merged for the reasons I outlined. If you really need the class to be this way, you will have to ship it with your library by implementing the interface yourself.

In summary: We won’t merge this. You are not blocked because you can implement the class in userspace. That’s it.

And now I am out here.

rela589n replied:
How would you do it if you your self needed it?

nicolas-grekas replied:
I think @xabbuh explained how already. E.g. copy/pasting the existing code and tweaking it on your side to fit your need.
I'm now locking this discussion because it's too time consuming and the matter has been handled on our side.

xabbuh reviewed (changes requested):
I don't see that these are valuable changes as they change the `ConstraintViolationBuilder` class in a way that makes it less obvious to use. Thus I am 👎 here.

xabbuh commented:
Re-reading the changes I am 👎 on merging these changes. I also doubt that the described use case is common that we need to provide a solution for it in core. You can easily ship a custom `ConstraintViolationBuilderInterface` implementation in your application/library if you see fit for it in you specific use case though.
